core/google/api/core/timeout.py

```diff
@@ -0,0 +1,210 @@
+            datetime_helpers.utcnow() +
+            datetime.timedelta(seconds=deadline))
+    else:
+        deadline = datetime.datetime.max
+
+    timeout = initial
+    while True:
+        now = datetime_helpers.utcnow()
+        yield min(
+            # The calculated timeout based on invocations.
+            timeout,
+            # The set maximum timeout.
+            maximum,
+            # The remaining time before the deadline is reached.
+            float((deadline - now).seconds))
+        timeout = timeout * multiplier
+
+
+@six.python_2_unicode_compatible
+class ExponentialTimeout(object):
+    """A decorator that adds a exponentially increasing timeout argument.
+
+    This is useful if a function is called multiple times. Each time the
+    function is called this decorator will calculate a new timeout parameter
+    based on the the number of times the function has been called.
+
+    For example
+
+    .. code-block:: python
+
+    Args:
+        initial (float): The initial timeout to pass.
+        maximum (float): The maximum timeout for any one call.
+        multiplier (float): The multiplier applied to the timeout for each
+            invocation.
+        deadline (float): The overall deadline across all invocations. This
+            is used to prevent a very large calculated timeout from pushing
+            the overall execution time over the deadline. This is especially
+            useful in conjuction with :mod:`google.api.core.retry`.
+    """
+    def __init__(
+            self,
+            initial=_DEFAULT_INITIAL_TIMEOUT,
+            maximum=_DEFAULT_MAXIMUM_TIMEOUT,
+            multiplier=_DEFAULT_TIMEOUT_MULTIPLIER,
+            deadline=_DEFAULT_DEADLINE):
+        self._initial = initial
+        self._maximum = maximum
+        self._multiplier = multiplier
+        self._deadline = deadline
+
+    def with_deadline(self, deadline):
+        """Return a copy of this teimout with the given deadline.
+
+        Args:
+            deadline (float): The overall deadline across all invocations.
+
+        Returns:
+            ExponentialTimeout: A new instance with the given deadline.
+        """
+        return ExponentialTimeout(
+            initial=self._initial,
+            maximum=self._maximum,
+            multiplier=self._multiplier,
+            deadline=deadline)
+
+    def __call__(self, func):
+        """Apply the timeout decorator.
+
+        Args:
+            func (Callable): The function to apply the timeout argument to.
+                This function must accept a timeout keyword argument.
+
+        Returns:
+            Callable: The wrapped function.
+        """
+        timeouts = _exponential_timeout_generator(
+            self._initial, self._maximum, self._multiplier, self._deadline)
+
+        @six.wraps(func)
+        def func_with_timeout(*args, **kwargs):
+            """Wrapped function that adds timeout."""
+            kwargs['timeout'] = next(timeouts)
```

Comment:
I expect the wrapped function to want `timeout` to be a `float` or coercible to a float. This must assume that the wrapped function will be OK receiving an iterator.